src/WebJobs.Extensions.DurableTask/Listener/DurableTaskTargetScaler.cs

```diff
@@ -38,46 +39,55 @@ public DurableTaskTargetScaler(string functionId, DurableTaskMetricsProvider met
-                $"maxConcurrentOrchestrators={this.MaxConcurrentOrchestrators}. maxConcurrentActivities={this.MaxConcurrentActivities}";
-
-            // target worker count should never be negative
-            if (numWorkersToRequest < 0)
+            DurableTaskTriggerMetrics? metrics = null;
+            try
             {
-                scaleControllerLog = "Tried to request a negative worker count." + scaleControllerLog;
-                this.logger.LogError(scaleControllerLog);
-
-                // Throw exception so ScaleController can handle the error.
-                throw new Exception(scaleControllerLog);
+                metrics = await this.metricsProvider.GetMetricsAsync();
+
+                // compute activityWorkers: the number of workers we need to process all activity messages
+                var workItemQueueLength = metrics.WorkItemQueueLength;
+                double activityWorkers = Math.Ceiling(workItemQueueLength / (double)this.MaxConcurrentActivities);
+
+                var serializedControlQueueLengths = metrics.ControlQueueLengths;
+                var controlQueueLengths = JsonConvert.DeserializeObject<IReadOnlyList<int>>(serializedControlQueueLengths);
+
+                var controlQueueMessages = controlQueueLengths.Sum();
+                var activeControlQueues = controlQueueLengths.Count(x => x > 0);
+
+                // compute orchestratorWorkers: the number of workers we need to process all orchestrator messages.
+                // We bound this result to be no larger than the partition count
+                var upperBoundControlWorkers = Math.Ceiling(controlQueueMessages / (double)this.MaxConcurrentOrchestrators);
+                var orchestratorWorkers = Math.Min(activeControlQueues, upperBoundControlWorkers);
+
+                int numWorkersToRequest = (int)Math.Max(activityWorkers, orchestratorWorkers);
+                this.scaleResult.TargetWorkerCount = numWorkersToRequest;
+
+                // When running on ScaleController V3, ILogger logs are forwarded to Kusto.
+                var metricsLog = $"Metrics: workItemQueueLength={workItemQueueLength}. controlQueueLengths={serializedControlQueueLengths}. " +
+                    $"maxConcurrentOrchestrators={this.MaxConcurrentOrchestrators}. maxConcurrentActivities={this.MaxConcurrentActivities}";
+                var scaleControllerLog = $"Target worker count for '{this.functionId}' is '{numWorkersToRequest}'. " +
+                    metricsLog;
+
+                // target worker count should never be negative
+                if (numWorkersToRequest < 0)
+                {
+                    scaleControllerLog = $"Error: target worker count for '{this.functionId}' was negative: '{this.functionId}'." +
+                        "An exception was thrown." + metricsLog;
+                    this.logger.LogError(scaleControllerLog);
+                    throw new Exception(scaleControllerLog);
```

Comment:
I'm just now noticing this, but it's bad practice to throw `Exception`. Consider using `InvalidOperationException` instead.

Reply:
Actually, doesn't this result in double-exception and double logging? If I'm reading the code correctly, we catch this exception a few lines down and make another call to `this.logger.LogError`. It seems like we need to reconsider the design here, because using exceptions as control flow like this is also bad practice.

Reply:
Yeah you're right, it's getting double logged; good call out.

A part of me wants a try-catch surrounding this whole operation so we can log any exceptions that arise, and to include in those logs the contextual information needed to debug this. That's what I do in the outer catch block.

Do you think it would suffice to remove the `LogError` when the target worker count is negative, and simply throw an exception? Alternatively, I could refactor the outer try-catch so that it ends right when the target worker count is calculated (i.e line 71)

Reply:
At a minimum we should remove the first `LogError` and rely on the one inside the catch-block. But I wonder if we even need that one since we're throwing. Wouldn't it be expected that some other code catches these exceptions and logs them as errors?

Reply:
@alrod, do you know what the expected behavior is for logging and throwing exceptions? Do we log the error or will it be handled in the scale controller?

Reply:
Still looking for an update on this.

Reply:
`numWorkersToRequest` is calcuated as:
`int numWorkersToRequest = (int)Math.Max(activityWorkers, orchestratorWorkers);`

How `numWorkersToRequest `can be negative?

If you want to throw an exception from `GetScaleResultAsync` I would just thorwing `InvalidOperationException/ArgumentOutOfRangeException` without writing the error log - Scale Controller will take care about this

Something like:
https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusTargetScaler.cs#L94

Reply:
Thanks @alrod. I agree that the current code cannot provide negative numbers, my check is just ensuring that remains true if the code is ever refactored. It may a bit too defensive, so I don't mind removing it if others think is overkill :).

@cgillum: I've removed the double logging on this commit: https://github.com/Azure/azure-functions-durable-extension/pull/2462/commits/05332642f6340b53cb6069c06eced3f7ed8e1273

Reply:
Changed the exception type to InvalidOperationException: https://github.com/Azure/azure-functions-durable-extension/pull/2462/commits/7f47657f53b5206168a4c92ad9181e8486f567b9